File: android/app/src/main/java/com/tubelimiter/app/limit/AlarmRules.kt
package com.tubelimiter.app.limit

/** Fire a "time is running out" nudge when the remaining time crosses each of these. */
val ALARM_MILESTONE_MINUTES = listOf(30, 10, 5, 1)

/**
 * 예약 차단 시작 몇 분 전에 알릴지. 확장 `lib/alarmRules.js`의 `SCHEDULE_SOON_LEAD_MINUTES`와
 * 같은 값이어야 하고, 알림 문구에도 이 값이 그대로 들어간다
 * (`R.plurals.nudge_schedule_upcoming`). 서비스 안에 숫자로 박아두면 한쪽만 바뀌어도 눈치채기
 * 어려워서 상수로 뺐다.
 */
const val SCHEDULE_SOON_LEAD_MINUTES = 10

data class AlarmState(
    val dateKey: String,
    val lastIntervalNotifyMillis: Long = 0L,
    val notifiedMilestones: Set<Int> = emptySet(),
)

/**
 * A nudge that is due, as data rather than text: the notification only exists on a device, but
 * the "did we already say this today" bookkeeping around it is what the unit tests exercise,
 * and unit tests cannot reach Android resources. The service turns these into strings
 * ([com.tubelimiter.app.service.UsageMonitorService.nudgeText]).
 */
sealed interface AlarmMessage {
    /** The recurring "you have been watching for a while" interval nudge. */
    data class WatchedMinutes(val minutes: Long) : AlarmMessage

    /** One of [ALARM_MILESTONE_MINUTES], crossed on the way down. */
    data class RemainingMinutes(val minutes: Int) : AlarmMessage

    /**
     * 예약 차단이 곧 시작된다. [minutes]에는 실제 남은 분이 아니라 예고 기준값
     * ([SCHEDULE_SOON_LEAD_MINUTES])이 담긴다 — 하루 한 번만 알리는데 7분 남았을 때 "7분 후"로
     * 바뀌면 규칙과 어긋나 보인다(확장 `evaluateAlarms`와 같은 판단).
     */
    data class ScheduleSoon(val minutes: Int) : AlarmMessage
}

data class AlarmOutcome(
    val state: AlarmState,
    val messages: List<AlarmMessage>,
) {
    val changed: Boolean get() = messages.isNotEmpty()
}

/**
 * Decides which nudges are due. Ported from the extension's `checkAlarms`; kept pure so
 * the "did we already say this today" bookkeeping is testable without a device.
 *
 * [minutesUntilScheduleStart]는 [minutesUntilNextScheduleStart]의 반환값(이미 활성 중이거나
 * 예약이 없으면 null)이고, [scheduleStartNotified]는 오늘 이미 예고했는지다. 예약 차단 예고
 * 판정을 여기 둔 건 확장 `evaluateAlarms`와 자리를 맞추기 위해서다 — 임계값과 판정이 서비스
 * 안에 흩어져 있으면 한쪽만 바뀌어도 드러나지 않는다. 저장 위치만 다르다: 확장은 이 표시를
 * `alarm_state`에 담지만 안드로이드는 별도 DataStore 키(`schedule_start_notified_date`)에
 * 날짜로 담으므로, 여기서는 읽기만 하고 갱신은 호출자가 한다.
 */
fun evaluateAlarms(
    previous: AlarmState?,
    todayKey: String,
    usedMillis: Long,
    limitMillis: Long,
    intervalMinutes: Int,
    milestonesEnabled: Boolean,
    minutesUntilScheduleStart: Long? = null,
    scheduleStartNotified: Boolean = false,
): AlarmOutcome {
    // A state from an earlier day carries stale bookkeeping, so start the day fresh.
    var state = previous?.takeIf { it.dateKey == todayKey } ?: AlarmState(todayKey)
    val messages = mutableListOf<AlarmMessage>()

    if (intervalMinutes > 0) {
        val intervalMillis = minutesToMillis(intervalMinutes)
        if (usedMillis - state.lastIntervalNotifyMillis >= intervalMillis) {
            // Snap to the interval boundary so a long gap does not queue up a burst.
            state = state.copy(lastIntervalNotifyMillis = (usedMillis / intervalMillis) * intervalMillis)
            messages += AlarmMessage.WatchedMinutes(usedMillis / 60_000)
        }
    }

    if (milestonesEnabled && !isUnlimited(limitMillis)) {
        val remaining = limitMillis - usedMillis
        val due = ALARM_MILESTONE_MINUTES.filter { minutes ->
            remaining > 0 &&
                remaining <= minutesToMillis(minutes) &&
                minutes !in state.notifiedMilestones
        }
        if (due.isNotEmpty()) {
            state = state.copy(notifiedMilestones = state.notifiedMilestones + due)
            due.forEach { messages += AlarmMessage.RemainingMinutes(it) }
        }
    }

    // 예약 차단 예고. 하루 한 번만 알리면 충분하므로 dedupe는 날짜로 한다 — 확장의
    // `state.scheduleStartNotified`와 같은 규칙이고, 안드로이드는 그 표시를 별도 키에 담는다.
    if (minutesUntilScheduleStart != null &&
        minutesUntilScheduleStart > 0 &&
        minutesUntilScheduleStart <= SCHEDULE_SOON_LEAD_MINUTES &&
        !scheduleStartNotified
    ) {
        messages += AlarmMessage.ScheduleSoon(SCHEDULE_SOON_LEAD_MINUTES)
    }

    return AlarmOutcome(state, messages)
}

File: extension/src/lib/alarmRules.js
// 알림 발화 판정 순수 로직. service-worker.js의 checkAlarms에서 "지금 무슨 알림을 띄워야
// 하는가 / 오늘 이미 띄웠는가"만 떼어냈다 — 실제 chrome.notifications 호출과 alarm_state
// 저장은 호출자 몫이다. 안드로이드 limit/AlarmRules.kt의 evaluateAlarms 대응이며,
// 한쪽 규칙을 고치면 다른 쪽도 맞춰야 한다.

/** 남은 시간이 이 값들을 밑돌 때마다 "얼마 안 남았다" 알림을 한 번씩 띄운다. */
export const ALARM_MILESTONE_MINUTES = [30, 10, 5, 1];

/** 예약 차단 시작 몇 분 전에 알릴지. 알림 문구에도 이 값이 그대로 들어간다. */
export const SCHEDULE_SOON_LEAD_MINUTES = 10;

/**
 * 알림 종류. 문구는 여기서 만들지 않는다 — 이 파일은 chrome.*를 모르는 순수 판정이고, 실제
 * 문구는 호출자(service-worker.js)가 chrome.i18n으로 붙인다(언어마다 어순이 달라 문자열을
 * 이어붙이는 대신 messages.json의 `$1` 자리에 분을 꽂아야 한다).
 */
export const ALARM_KIND = Object.freeze({
  interval: 'interval',
  milestone: 'milestone',
  scheduleSoon: 'scheduleSoon'
});

/** 오늘치 알림 장부의 초기값. 날짜가 바뀌면 이걸로 새로 시작한다. */
export function createAlarmState(date) {
  return { date, lastIntervalNotifyMs: 0, notifiedMilestones: [], scheduleStartNotified: false };
}

// 어제 장부를 그대로 쓰면 "이미 알렸음" 표시가 남아 오늘 알림이 통째로 죽는다.
// 날짜가 다르면 버리고, 같으면 모르는 필드까지 보존한 채 복사해서 쓴다.
function restoreState(previous, date) {
  if (!previous || previous.date !== date) return createAlarmState(date);
  return {
    ...previous,
    lastIntervalNotifyMs: previous.lastIntervalNotifyMs || 0,
    notifiedMilestones: Array.isArray(previous.notifiedMilestones) ? [...previous.notifiedMilestones] : [],
    scheduleStartNotified: !!previous.scheduleStartNotified
  };
}

/**
 * 이번 틱에 띄워야 할 알림과 갱신된 장부를 계산한다.
 *
 * @param {object|null} previous 저장돼 있던 alarm_state
 * @param {object} inputs
 * @param {string} inputs.date 오늘 날짜(4시 컷오프 기준, getTodayDate())
 * @param {number} inputs.usedMs 오늘 사용량(다른 기기 몫 합산 후)
 * @param {number} inputs.limitMs 오늘 한도 (무제한이면 Infinity — 마일스톤은 조용해진다)
 * @param {number} inputs.intervalMinutes N분 주기 알림 간격 (0이면 끔)
 * @param {boolean} inputs.milestonesEnabled 남은 시간 마일스톤 알림 사용 여부
 * @param {number|null} inputs.minutesUntilScheduleStart 다음 예약 차단까지 남은 분
 *   (minutesUntilNextScheduleStart의 반환값 — 이미 활성 중이거나 예약이 없으면 null)
 * @returns {{state: object, changed: boolean, notifications: Array<{kind: string, minutes: number}>}}
 *   notifications는 발화 순서대로이고, 각 항목은 종류(ALARM_KIND)와 문구에 들어갈 분(minutes)만
 *   담는다. 문구도 알림 id도 호출자가 붙인다 — 주기/예약 알림은 매번 고유한 id여야 하는데
 *   (같은 id면 크롬이 배너를 다시 안 띄우고 조용히 업데이트만 한다) 그건 Date.now()가 필요해
 *   순수 함수 밖의 일이다.
 */
export function evaluateAlarms(previous, {
  date,
  usedMs = 0,
  limitMs = Infinity,
  intervalMinutes = 0,
  milestonesEnabled = true,
  minutesUntilScheduleStart = null
} = {}) {
  const state = restoreState(previous, date);
  const notifications = [];
  let changed = false;

  if (intervalMinutes > 0) {
    const intervalMs = intervalMinutes * 60 * 1000;
    if (usedMs - state.lastIntervalNotifyMs >= intervalMs) {
      // 다음 기준점을 지난 경계에 스냅시킨다. 오랫동안 못 돌았어도 밀린 알림이
      // 한꺼번에 쏟아지지 않고 이번 한 번으로 끝난다.
      state.lastIntervalNotifyMs = Math.floor(usedMs / intervalMs) * intervalMs;
      changed = true;
      notifications.push({ kind: ALARM_KIND.interval, minutes: Math.floor(usedMs / 60000) });
    }
  }

  // 한도가 무제한(Infinity)이면 "남은 시간"이라는 개념 자체가 없으므로 건너뛴다.
  if (milestonesEnabled && Number.isFinite(limitMs)) {
    const remaining = limitMs - usedMs;
    for (const minutes of ALARM_MILESTONE_MINUTES) {
      // remaining > 0 조건 때문에 한도를 이미 넘긴 뒤엔 조용하다 (그땐 차단 화면이 뜬다).
      if (remaining > 0 && remaining <= minutes * 60 * 1000 && !state.notifiedMilestones.includes(minutes)) {
        state.notifiedMilestones.push(minutes);
        changed = true;
        notifications.push({ kind: ALARM_KIND.milestone, minutes });
      }
    }
  }

  // 예약 차단 시작 10분 전 알림. 하루 한 번만 알리면 충분하므로 날짜별로 한 번만 dedupe한다.
  // minutes에는 실제 남은 분이 아니라 예고 기준값(10)을 담는다 — 문구가 "10분 후"로 고정이라
  // 7분 남았을 때 "7분 후"로 바뀌면 하루 한 번만 알린다는 규칙과 어긋나 보인다.
  if (minutesUntilScheduleStart !== null && minutesUntilScheduleStart > 0 &&
      minutesUntilScheduleStart <= SCHEDULE_SOON_LEAD_MINUTES && !state.scheduleStartNotified) {
    state.scheduleStartNotified = true;
    changed = true;
    notifications.push({ kind: ALARM_KIND.scheduleSoon, minutes: SCHEDULE_SOON_LEAD_MINUTES });
  }

  return { state, changed, notifications };
}
